### ISSUE TITLE
Can't send any POST body with request/node

### ISSUE THREAD

**Issue by sebilasse** [dojo/core#197]
in `request/node` : 

requestOptions SHOULD have at least `body` :

``` typescript
interface HttpsOptions extends Options {
    body?: any;
    // ...
}
```

and

``` typescript
const requestOptions: HttpsOptions = {
    agent: options.agent,
    auth: parsedUrl.auth || options.auth,
    body: options.data,
    // ...
}
```

additionally I think there are two more options when sending POST data in nodeJS :
`form: any;`
and
`json: boolean;`

**Comment by dylans** [comment]
@sebilasse thanks, that seems reasonable. Do you intend to create a pull request, or should we handle that?

**Comment by sebilasse** [comment]
@dylans Will investigate the next days (not sure about the `json` parameter, poor node documentation on that). Might be helpful to see how https://github.com/request/request deals with it. 
However, created PR :

**Comment by sebilasse** [comment]
some more findings, all about `request/node`
It is also important to set the headers as lowercase 

``` js
if (typeof options.headers === 'object') {
    options.headers = Object.keys(options.headers).reduce(function(o, key) {
        o[key.toLowerCase()] = options.headers[key];
        return o;
    }, {});
}
options = mixin({maxRedirects: 5}, options);
options.headers = Object.keys(options.headers).reduce(function(o, key) {
    o[key.toLowerCase()] = options.headers[key];
    return o;
}, {});
```

if you e.g. just check 
`if (!('user-agent' in requestOptions.headers)) {`

What I also stumbled upon is this note in `request/node` :
_"TODO: This redirect code is not 100% correct according to the RFC; needs to handle redirect loops and restrict/modify certain redirects"_
This feels pretty important and could also be a (trackable) issue (especially because redirect loops will occur). Apart from that I guess that if an options.headers.host is already set it MUST be rewritten by the redirect code (think of `www.` redirects) :

``` js
    if (options.headers && options.headers['host'] && requestOptions.host) {
        options.headers.host = requestOptions['host'];
    }
```

My temporary fix to prevent infinite loops in node is to send a `redirectCount` with the options which can be limited by `maxRedirects` and never allows more than 5 –

for now see this [GIST](https://gist.github.com/sebilasse/e54fbd7722c98de5ad1e7dd8ab91f52d) for the complete file, modifications are marked by "TODO MOD"

**Comment by rorticus** [comment]
As far as I can tell, sending POST data is working right now. You can send data with `request/node` using the existing `data` option. It’s called `data` to make it consistent with `request/xhr` .

I just ran this,

``` js
import request from './request/node';

request('<requestb.in url here>', {
    method: 'POST',
    data: 'test'
}).then(() => {
    console.log('all done');
});
```

and the data is POSTed successfully. 

In regards to the comments about header case insensitivity and redirect TODO, see the following PRs,

[Case insensitivity when checking for user-agent header, fixes #168 by rorticus · Pull Request #236 · dojo/core · GitHub](https://github.com/dojo/core/pull/236)

[Making redirects follow (more closely anyway) the RFC, fixes #207 by rorticus · Pull Request #212 · dojo/core · GitHub](https://github.com/dojo/core/pull/212)

Unless we want to add convenience parameters for automatically encoding form/json data, I don’t think there is anything actionable here anymore.

**Comment by rorticus** [comment]
Closing for now. If you think this is still an issue, please re-open!
